[SDFAB-1007] Drop CLI interactive mode in favor of gRPC and Protobuf (#18)
omec-project/pfcpsim: README.md

# pfcpsim

> pfcpsim is a simulator to interact with PFCP agents. Can be used to emulate a 4G SGW / 5G UPF.

Reviewer: ```suggestion
pfcpsim is a simulator to interact with PFCP agents. Can be used to simulate a 4G SGW-C / 5G SMF.
```

## Interaction

> pfcpsim is designed to work within a containerized environment. The docker image comes with both client (`pcpfsimctl`) and server (`pfcpsim`).

Reviewer: ```suggestion
pfcpsim is designed to work within a containerized environment. The docker image comes with both client (`pfcpctl`) and server (`pfcpsim`).
```

`PFCPClient` is embedded in a gRPC Server. Interaction between pfcpsim and pfcpsimctl is performed through RPC calls, as shown in the following schema: 

![Alt text](docs/images/schema.svg?raw=true "schema")

> ## Example

Reviewer: ```suggestion
## Getting started
```

> In the following example we will borrow `mock-up4` environment, from [upf-epc repository](https://github.com/omec-project/upf-epc/tree/master/test/integration).

Reviewer: I guess we can present the steps to follow, without any PFCP Server.

Using mock-up4's `docker-compose.yml` will create a pfcp agent container listening on `0.0.0.0:8805`:

1. Build the container locally (You can skip this once docker image is published by the CI):
```bash
make build-pfcpsim
```
2. Create the container using host network (to be able to connect to 127.0.0.1:8805)
```bash

> docker container run --rm -d --name pfcpsim --net host pfcpsim:0.1.0-dev

Reviewer: How to we pass the address of a PFCP server?

Author: You can pass the address of a PFCP server using the `Configure` RPC. I will explain this better

```

> 3. Use pfcpsimctl by using `docker exec` on the same container to configure the server (not passing any flag will use mock-up4's default values)

Reviewer: ```suggestion
3. Use `pfcpctl` to configure the server (not passing any flag will use mock-up4's default values)
```

Reviewer: what are the mock-up4's default values? We shouldn't use names (mock-up4) specific to other projects (upf-epc in this case).

```bash

> docker exec pfcpsim /bin/pfcpctl -c configure

Reviewer: I'm wondering if we can make `pfcpctl` a globally exported binary, so that we don't need `/bin` prefix.

Author: Done

```

4. Associate
```bash
docker exec pfcpsim /bin/pfcpctl -c associate
```

5. Create 5 sessions
```bash

> docker exec pfcpsim /bin/pfcpctl -c create --count 5 --baseID 2 --ue-pool 17.0.0.0/24 --gnb-addr 198.18.0.10

Reviewer: ```suggestion
docker exec pfcpsim /bin/pfcpctl -c create --count 5 --baseID 2 --ue-pool 17.0.0.0/24 --nb-addr 198.18.0.10
```

`nb-addr` works fine for both 4G and 5G

```

6. Delete the sessions
```bash
docker exec pfcpsim /bin/pfcpctl -c delete --count 5 --baseID 2
```

> 7. Disconnect from remote peer

Reviewer: Does `interrupt` actually disconnects from a remote peer? Or does it just kills the `pfcpsim` ?

Reviewer: Should we use `disassociate` command for disconnection?

Author: Done. Deleted the `Interrupt` RPC. `Disassociate` will now perform disassociation along with disconnection from remote peer.

> # pfcpsim

Author: Done. I plan on improving the README once we publish the pfcpsim's docker image
